tests/framework/prototype.py

```diff
@@ -0,0 +1,279 @@
+import yaml
+from typing import Optional
+from kubernetes import client, config, utils
+import os
+import time
+import docker
+import logging
+import unittest
+import jsonpatch
+
+# Utility functions
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+def search_path(cr, steps):
+    curr = cr
+    for step in steps:
+        if step.isnumeric():
+            int_step = int(step)
+            if int_step >= len(curr) or int_step < 0:
+                return None
+            curr = curr[int(step)]
+        elif step in curr:
+            curr = curr[step]
+        else:
+            return None
+    return curr
+
+'''
+Functions for cluster preparation. Typical workflow:
+  Delete KinD cluster -> Create KinD cluster -> Install CRD -> Download Images (from DockerHub) ->
+  Load images into KinD cluster -> Install KubeRay operator
+'''
+def delete_kind_cluster() -> None:
+    """Delete a KinD cluster"""
+    os.system("kind delete cluster")
+
+def create_kind_cluster():
+    """Create a KinD cluster"""
+    os.system("kind create cluster")
+    os.system("kubectl wait --for=condition=ready pod -n kube-system --all --timeout=900s")
+
+def install_crd():
+    """Install Custom Resource Definition (CRD)"""
+    KUBERAY_VERSION = "v0.3.0"
+    os.system("kubectl create -k \"github.com/ray-project/kuberay/manifests/cluster-scope-resources?" +
+              "ref={}&timeout=90s\"".format(KUBERAY_VERSION))
+
+def download_images(images):
+    """Download Docker images from DockerHub"""
+    docker_client = docker.from_env()
+    for image in images:
+        docker_client.images.pull(image)
+    docker_client.close()
+
+def kind_load_images(images):
+    """Load downloaded images into KinD cluster"""
+    for image in images:
+        os.system(f'kind load docker-image {image}')
+
+def install_kuberay_operator():
+    KUBERAY_VERSION = "v0.3.0"
+    os.system("kubectl apply -k \"github.com/ray-project/kuberay/manifests/base?" +
+              "ref={}&timeout=90s\"".format(KUBERAY_VERSION))
+
+def check_cluster_exist():
+    return os.system("kubectl cluster-info --context kind-kind") == 0 
+
+'''
+Configuration Test Framework Abstractions: (1) Mutator (2) Rule (3) RuleSet (4) CREvent
+'''
+
+# Mutator: Mutator will start to mutate from `baseCR`. `patch_list` is a list of JsonPatch, and you can
+#          specify multiple fields that want to mutate in a single JsonPatch.
+class Mutator:
```

Comment:
I'm thinking about ways in which we can simplify things -- do we actually need the Mutator class at this point?
At this point, we can simply specify some json patches and apply them to generate CRs. It's not clear to me that we need an extra class to do this.

Reply:
+1. This class can be a helper function

Reply:
The intention, as described in the design doc, is to support, via inheritance, different ways of applying json patches to create collections of CRs.
This is all good. Tough, noting again my wariness of premature abstraction.